### ISSUE TITLE
Support spawnUri to start a "strong" isolate from a "weak" process

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#40632]
We want to be able to run package tests – all using `package:test` – with null safe strong semantics ASAP.

This means we need to migrate the target package AND all of the test infrastructure code that runs with it. (Think the `test` API – `expect`, etc)

If the granularity of "strong" is a process, then we need to migrate all of `package:test` and it's dependencies before we can run tests.

**52 total packages**

If we can run the test process in weak mode and then run individual test isolates in strong mode the number of packages drops to...

**12 total packages**

### Question 1: *can* we run a "strong" isolate in a "weak" process?

I was told this is a CFE question. @johnniwinther ?

### Question 2: *how* do we spawn a "strong" isolate in a "weak" process?

Options
1) Detect `// @dart = 2.8`comment and "just do it"
  - very simple
  - potential for launching code "strong" when it's not ready...hrm...
 
2) Add an argument to `Isolate.spawnUri`
  - `bool spawnStrongNullSafe` or similar
  - Not sure if we want another API here. Best to avoid.
  - Thoughts @lrhn ?

3) Create another magic comment syntax
  - `// @dart_run_null_safe_strong` or...something
  - inventing something new
  - no new API needed
  - Low chance of accidentally running something existing accidentally in strong mode – we can document it little or not at all. Just wire `pkg:test` to use it when creating the `data:` URI containing the test wrapper.

CC @vsmenon

**Comment by kevmoo** [comment]
CC @a-siva

**Comment by kevmoo** [comment]
@a-siva did you say all of this work is done now?

CC @vsmenon

**Comment by a-siva** [comment]
Which of the options listed above did we agree upon?

Option 1 - (doesn't work for the 'pub test' scenario I guess)
Option 2 - would work, it will be a breaking change, we would need to roll that out soon if we are going with that.
Option 3 -  Not sure about this option ??

**Comment by a-siva** [comment]
Another potential option is to overload the parameter 'checked' to also mean null-safety when --enable-experiment=non-nullable is set.

**Comment by leafpetersen** [comment]
I don't know any details about how spawnURI works, but assuming that we disallow running mixed mode code in strong mode, is the following an option:
  - If the code is mixed mode, spawn a weak isolate
  - If the code is all opted in, spawn a strong isolate

It feels like this might be implementable?  A naive implementation would be to literally spawn a strong isolate, watch for the compile error that tells you you've tried to compile mixed code in strong mode, and if so, re-spawn as a weak isolate?  

This wouldn't allow you to spawn a weak isolate to run unmixed code, but we could probably live without that?

**Comment by lrhn** [comment]
Ad 2.
Adding a parameter to `spawnUri` means that we have to remove it again later when we stop supporting weak mode. That will likely mean deprecating it for half a year or so, and then removing it entirely, breaking all code that still uses the flag. Or more likely, just removing it in Dart 3.0.

The flag will need to keep existing as long as we do support weak mode, so that's after releasing Null Safety. So, using the checked flag with the experiment flag doesn't work.

It's a static method, so there are not inheritance issues with changing it. Adding an optional parameter is not a breaking change.

Ad 1. This would enable strong mode for all Null Safety enabled libraries? How would it work after we release 2.10?
Ad 3. I don't see this working if you need to spawn code that you are not in control of.

I think Leaf's idea is the most promising.
There is no weak/strong mode switch. You get strong mode if your entire program is Null Safe enabled, otherwise you get weak mode. There is nothing you can do to prevent that except add a non-Null Safe library to the program.

**Comment by eernstg** [comment]
I think we may certainly want to have a weak/strong mode switch: It's going to be highly confusing for developers if the different semantics kicks in (e.g., a bunch of `if (x is T) ..` locations suddenly choose the other branch!) just because the last piece of legacy code in some indirect import of an import has been migrated.

- If the code is all opted in, and full null-safety is enabled, spawn a strong isolate

**Comment by natebosch** [comment]
> when we stop supporting weak mode. That will likely mean deprecating it for half a year or so, and then removing it entirely, breaking all code that still uses the flag. Or more likely, just removing it in Dart 3.0.

Are we going to stop supporting weak mode sometime during Dart `2.x`? I had thought that disabling weak  mode was the trigger for us to bump the major version.

> It's going to be highly confusing for developers if the different semantics kicks in (e.g., a bunch of `if (x is T) ..` locations suddenly choose the other branch!) just because the last piece of legacy code in some indirect import of an import has been migrated.

If someone performed a migration that only works in weak mode - they did that migration wrong. I think it's OK to break them. To opt back out all they need to do is tag the "entrypoint" library with `// @dart=2.7` to get back to weak mode.

> * If the code is all opted in, and full null-safety is enabled, spawn a strong isolate

What does it mean in this context for  full null-safety to be enabled? Do you mean on the starting Isolate in the VM? The user likely does not have control over that in the test case.

**Comment by leafpetersen** [comment]
Another thought: can we encode this in the URI in some way?  e.g. as a parameter in the URI?

**Comment by kevmoo** [comment]
> Another thought: can we encode this in the URI in some way? e.g. as a parameter in the URI?

Uh. URI in general, sure. Use an anchor tag or similar.

For a `file:///` URI, I don't think there is a lot of room for interpretation.

The test package does use a data URI to spawn things, though. Maybe there's room there to encode something?

https://github.com/dart-lang/test/blob/c0c7761c186cd1e17a4d0694059d467d86415c7b/pkgs/test_core/lib/src/util/dart.dart#L21-L29

What do "parameters" do in Data URIs? https://api.dart.dev/stable/2.7.1/dart-core/Uri/Uri.dataFromString.html

**Comment by lrhn** [comment]
> What do "parameters" do in Data URIs?

Whatever the media-type intends them to do. They are *media type* parameters, so the classical use is `text/plain;charset=iso-8859-1`. You can add more parameters, but it might confuse code which thinks it recognizes the media.

If we use our own media-type, like "application/x.dart;nnbd=weak",  then we'd be in charge of interpreting it (and nothing else would understand it at all). It does suggest some level of specification, and since it's for temporary use, it might just be adding unnecessary complexity in the longer run. It also suggests that an HTTP server could send the same media type in its HTTP header, and we should probably handle it there too, not just in data URIs. And it doesn't work for `file:` URIs unless the file system somehow supports metadata for files.

(We should register `application/dart` as a media type).

I'm not sure I *like* using data-URI parameters for specifying *isolate* parameters. The data URI represents a single file, not the isolate, so it feels like the wrong place to put it.

Another issue is that with modular AoT compilation, we need to know at compile-time of the isolate code (by itself) whether code is going to be weak or strong. 
That means that we can't really specify it in the spawning code. Not for AoT. (Do isolates work in AoT at all?)
That suggests that it's a build option, not a code choice, how an isolate is compiled.

On the other hand, the stand-alone VM will compile at run-time, so we need some way to pass the flag there. 

So, we need both?
That suggests that what the code can pass is at most a request, not a requirement. And it's something that we are going to ignore at some point in the future and always choose strong.

We could make the default be "strong if all code is Null Safe, weak otherwise", with a way to override in either direction. If we don't want to introduce temporary API, we can cheat and use zone variables:
```dart
var nnbdWeak = true == (zone.current[#IsolateSpawnUriWeak] ?? (!source.isAllNullSafe));
```
Then code needing to force the choice can run `Isolate.spawnUri` in a zone overriding that variable.
When we need it to stop working, we'll just stop checking the zone variable.

**Comment by kevmoo** [comment]
> > What do "parameters" do in Data URIs?
> 
> Whatever the media-type intends them to do. They are _media type_ parameters, so the classical use is `text/plain;charset=iso-8859-1`. You can add more parameters, but it might confuse code which thinks it recognizes the media.
> 
> If we use our own media-type, like "application/x.dart;nnbd=weak", then we'd be in charge of interpreting it (and nothing else would understand it at all). It does suggest some level of specification, and since it's for temporary use, it might just be adding unnecessary complexity in the longer run. It also suggests that an HTTP server could send the same media type in its HTTP header, and we should probably handle it there too, not just in data URIs. And it doesn't work for `file:` URIs unless the file system somehow supports metadata for files.
> 
> (We should register `application/dart` as a media type).
> 
> I'm not sure I _like_ using data-URI parameters for specifying _isolate_ parameters. The data URI represents a single file, not the isolate, so it feels like the wrong place to put it.

Shucks. Just spit-balling.

> Another issue is that with modular AoT compilation, we need to know at compile-time of the isolate code (by itself) whether code is going to be weak or strong.
> That means that we can't really specify it in the spawning code. Not for AoT. (Do isolates work in AoT at all?)
> That suggests that it's a build option, not a code choice, how an isolate is compiled.
> 
> On the other hand, the stand-alone VM will compile at run-time, so we need some way to pass the flag there.
> 
> So, we need both?
> That suggests that what the code can pass is at most a request, not a requirement. And it's something that we are going to ignore at some point in the future and always choose strong.
> 
> We could make the default be "strong if all code is Null Safe, weak otherwise", with a way to override in either direction. If we don't want to introduce temporary API, we can cheat and use zone variables:

I thought we couldn't do this. At least @a-siva made it sound that way. Or we try strong. If it fails, we fall-back to weak.

> ```dart
> var nnbdWeak = true == (zone.current[#IsolateSpawnUriWeak] ?? (!source.isAllNullSafe));
> ```
> 
> Then code needing to force the choice can run `Isolate.spawnUri` in a zone overriding that variable.
> When we need it to stop working, we'll just stop checking the zone variable.

Maybe? 🤷‍♂

**Comment by leafpetersen** [comment]
> Another issue is that with modular AoT compilation, we need to know at compile-time of the isolate code (by itself) whether code is going to be weak or strong.
> That means that we can't really specify it in the spawning code. Not for AoT. (Do isolates work in AoT at all?)

I don't believe that spawnUri works with AOT.

> On the other hand, the stand-alone VM will compile at run-time, so we need some way to pass the flag there.

I think this is the only use case we need to solve.

> We could make the default be "strong if all code is Null Safe, weak otherwise",

This is my proposal above.  @a-siva has indicated that from the VM standpoint, it's probably feasible to delay the decision as to whether to run in strong or weak mode until after the CFE has been called out to for compilation.  But I believe that @johnniwinther wasn't sure that it was feasible for the CFE to do this efficiently, since it generally wants to know what mode it is compiling in.  I'm pretty sure the "try to compile in strong mode and then retry if you error out because you hit opted out code" approach could be made to work, but I would worry about adding a lot of latency to our legacy mode startup.


> If we don't want to introduce temporary API, we can cheat and use zone variables:
> 
> ```dart
> var nnbdWeak = true == (zone.current[#IsolateSpawnUriWeak] ?? (!source.isAllNullSafe));
> ```
> 
> Then code needing to force the choice can run `Isolate.spawnUri` in a zone overriding that variable.
> When we need it to stop working, we'll just stop checking the zone variable.

@jakemac53 @natebosch What do you think about this approach?  I think we're find with this being unadvertised API only used in our own packages, right?

**Comment by jakemac53** [comment]
I don't see a problem with using an unadvertised api like this, but I do think we would essentially be stuck with it until Dart 3.0 regardless of how "public" it is.

There will be a version of package:test out there forever which will be trying to use it, and the upper bound on the sdk will ideally be <3.0.0 since we have other versions which already have that constraint.

One alternative could be to only use this option only in some git branch of package:test, but that means anybody wanting to run in strong mode would have to have a git override.

**Comment by lrhn** [comment]
I expect any weak/strong mode toggle to be available until Dart 3.0.0, and obviously stop working there.

Whether making that an API change at 3.0.0, or hacking around that with a zone variable that just stops having any effect, is just a question of how invasive we want the change to be.

Dart 3.0.0 is going to be massively breaking to any code that isn't running Null Safe Dart yet, and a complete non-event to code that is. At that point, you should be running your tests in strong mode anyway. The only question is whether you are also running them in weak mode, but the fix to that is to stop trying when your package expects Dart 3.0.0.

**Comment by kevmoo** [comment]
We only need the work around until all of pkg:test is migrated, right? Then
we don't care about per-isoalet.

**Comment by jakemac53** [comment]
> We only need the work around until all of pkg:test is migrated, right? Then we don't care about per-isoalet.

package:test and all 52 of its deps

If we publish a version of package:test which relies on this and doesn't have a tight sdk constraint then we need it until Dart 3.0, and a tight sdk constraint would have other issues.

**Comment by natebosch** [comment]
> Then we don't care

I suspect we do care - it will have an impact on UX. If we support per-isolate strength then we can run all of a test suite in one go, and either detect or allow configuration per file. If we don't support it then the user will manually need to divide their tests into strong and weak, and invoke the test runner twice.

**Comment by kevmoo** [comment]
We'll want to get all 52 deps migrated sooner rather than later, right?

**Comment by natebosch** [comment]
> We'll want to get all 52 deps migrated sooner rather than later, right?

Yes, but I don't know if that changes the conversation much.

**Comment by a-siva** [comment]
After some more analysis, I feel an easy way forward would be to introduce a command line option in the VM '--spawn-isolates-in-strong-mode'. This option would be off by default and when specified all isolates spawned using SpawnUri would be started in strong mode if '--enable-experiment=non-nullable' is specified.
When '--enable-experiment=non-nullable' is not specified the option '--spawn-isolates-in-strong-mode' would be an error.

The 'pub' tool could be wired such that it invokes the VM with this command line option.

**Comment by kevmoo** [comment]
I think this will work for pkg:test. Please give us a CL ASAP so we can try it out!

**Comment by natebosch** [comment]
> The 'pub' tool could be wired such that it invokes the VM with this command line option.

If we can't have it be automatic based on the migration status of the sources I'd strongly prefer if it could be handled by the test runner and not the pub tool. The UX of `DART_VM_OPTIONS=--spawn-isolates-in-strong-mode pub run test` is not great. We need to support either strong or weak mode tests for the foreseeable future.

Is it possible we can pick based on the migration status of the library with `main`?

**Comment by jakemac53** [comment]
Afaik this option is not actually feasible, the problem is that `pub` itself also spawns an isolate for the test runner, which needs to be in weak mode (it won't be fully migrated). The test runner isolate then needs to spawn a strong isolate for each test (that is opted in ideally).

So we always have a situation where we need isolates spawned in multiple modes, at least until all of package:test and its transitive deps are migrated.

**Comment by natebosch** [comment]
> at least until all of package:test and its transitive deps are migrated.

I think we could get past this hurdle - I worry about the UX for people doing their own migrations more.

**Comment by jakemac53** [comment]
> I think we could get past this hurdle - I worry about the UX for people doing their own migrations more.

This would mean for all 50+ packages we can't run `pub run test`. I think that is a pretty big problem.

**Comment by natebosch** [comment]
@grouma @jakemac53 and I had a discussion about this. I feel more strongly than ever that choosing this based on the migration status on the entrypoint file is the right choice.

- We can make this consistent across all surfaces - `build_web_compilers` and `build_vm_compilers` can use this when compiling apps and test. `package:test` can use this for tests, both VM and dart2js. `dart` command line can use this for direct invocations, or AOT compiles, as well as spawned isolates from Dart source.
- The UX feels the same everywhere. You don't get surprised when the flag looks one way when invoking `dart`, a different way for `pub run`, a different way for `pub run test`. Etc. The opt _out_ flags, where we add them, might be a little different.
- Even if we fail to add an opt out flag somewhere because we overlook something, no user will be hard stuck. There is _always_ the option of writing a small file with the `// @dart=2.7`, import the other file and run its `main`. 
- Users are not stuck passing flags to get safety on Travis. You won't get into a situation where you were passing `--null-safety` locally during the migration, but forget to update `.travis.yaml` and don't realize you lose out on safety in the long term.
- There is no churn when we hit Dart 3 and all the opt-in flags become tech debt that should (eventually? immediately?) be removed.

**Comment by johnniwinther** [comment]
Note that from the CFE perspective:
1. A non-trivial amount of work has been done at the point in time when the language-version of 
the entry-point has been computed. By the current architecture this means that the only way to ensure consistency within the CFE about strong/weak mode is to assume one mode and then completely restart the CFE when that assumption turned out to be false.
2. The notion of "one entrypoint" does not exist for all uses of the CFE (e.g. modular compilation and platform compilation). We need a different strategy for this cases.
3. When compiling (partially) from .dill, the mode has already been baked into the .dill and the entry point + additional .dill files must match for the compilation to work. In this case restarting with a different mode might not work since the compilation might need different additional .dill files for this mode.

**Comment by jakemac53** [comment]
Ya I think for the compilers it makes sense they should be given the mode ahead of time as an argument. The tools that need to do entrypoint detection should be the tools that invoke the compilers (build_web_compilers, test, vm, blaze)

**Comment by johnniwinther** [comment]
And a weird case: If you in hot reload change the `// @dart=` annotation in the entry-point, should the mode change mid-flight?

**Comment by jakemac53** [comment]
> And a weird case: If you in hot reload change the `// @dart=` annotation in the entry-point, should the mode change mid-flight?

It should recompile from scratch imo.

**Comment by natebosch** [comment]
> 1. A non-trivial amount of work has been done at the point in time when the language-version of
>    the entry-point has been computed. By the current architecture this means that the only way to ensure consistency within the CFE about strong/weak mode is to assume one mode and then completely restart the CFE when that assumption turned out to be false.

That's one of the things I was hoping to frontload - for the "about to run" case how much work would it be to refactor so that detection happens first? Or how difficult will it be to have the VM discover this before calling the CFE, and how much perf hit will that be?

> 2. The notion of "one entrypoint" does not exist for all uses of the CFE (e.g. modular compilation and platform compilation). We need a different strategy for this cases.
> 3. When compiling (partially) from .dill, the mode has already been baked into the .dill and the entry point + additional .dill files must match for the compilation to work. In this case restarting with a different mode might not work since the compilation might need different additional .dill files for this mode.

Yes. We'll be doubling the number of "Platforms" in build systems. Strong DDC and weak DDC will be separate pipelines. We don't think that this issue changes that at all. When we compile the entrypoint we'll use the same migrated/not migrated criteria to decide which tree of dependencies to pull in, strong or weak.




> > And a weird case: If you in hot reload change the `// @dart=` annotation in the entry-point, should the mode change mid-flight?
> 
> It should recompile from scratch imo.

+1 on a full recompile. We should treat that change like any other change which we can't support for a hot reload.
Realistically, any change in the entrypoint should be a restart instead of a hot reload. We hope that there is little in that file other than `main`, and we can't hot reload a change in `main` anyway right?

**Comment by kevmoo** [comment]
RE "detection happens first" – we really only care about `package_conifg.json` + any `// @dart` flag.

**Comment by a-siva** [comment]
Can one of you try the CL https://dart-review.googlesource.com/c/sdk/+/140882 to see if it works for the pub use case. The flag is "--spawnuri_with_null_safety"

**Comment by jakemac53** [comment]
It may have gotten buried in other comments but this approach would only work for running a pub binary (ie: the test runner itself) in strong mode, but what we need is the test runner in weak mode and then the tests that spawns in strong mode.

We have 3 layers of isolates:

- Root isolate (weak)
  - Test runner isolate (weak)
    - Actual test isolates (strong)

**Comment by kevmoo** [comment]
@johnniwinther – any thoughts on what @natebosch had to say above? This is truely blocking!

**Comment by a-siva** [comment]
Is it a lot of work to migrate the test runner to be fully null safe so it can run in strong mode?

**Comment by natebosch** [comment]
> Is it a lot of work to migrate the test runner to be fully null safe so it can run in strong mode?

Yes. Around 50 packages, and having to migrate those without being able to use the test runner will make it even more difficult.

If it's not feasible to detect migration status of the entrypoint file in the CFE, can the VM do it before calling the CFE?

**Comment by natebosch** [comment]
To reiterate - we also believe that detecting status based on the entrypoint library will be a significantly better long term UX for our users than a VM flag.

**Comment by jakemac53** [comment]
Right, the idea that running in strong mode will require a flag until Dart 3.0 would be very sad 😢.

This really needs to be automatically inferred one way or another, and I think using the entrypoint file language version is the least worst option that has been proposed. It can be computed by reading only two files (package_config.json and the entrypoint file), and it should in most cases do what users expect.

**Comment by natebosch** [comment]
For the case of `data:` URIs in `spawnUri` there may not be a perfect answer about what language version to use. I think we would be fine with any of:

- requiring a `// @dart=` comment in order to get a newer version than `2.7` or some default.
- picking the default version based on the `package:` of the library spawning the URI.
- Some other scheme that works for you - so long as we can always include a `// @dart=` comment to pick it explicitly.

For the test runner use case we expect we'll need to do our own language version detection and we can custom construct the `data:` URI of the library we use to spawn the isolate.

**Comment by a-siva** [comment]
> Ya I think for the compilers it makes sense they should be given the mode ahead of time as an argument. The tools that need to do entrypoint detection should be the tools that invoke the compilers (build_web_compilers, test, vm, blaze)

I am not sure I understand how the tools detect the mode in order to pass it down to CFE.

More I think about this and given the new requirement that there is an additional isolate (test runner) that needs to run in weak mode, I feel adding an additional parameter to the spawnUri call to specify the mode in which the isolate needs to run is the cleanest from the VM perspective.

**Comment by jakemac53** [comment]
> I feel adding an additional parameter to the spawnUri call to specify the mode in which the isolate needs to run is the cleanest from the VM perspective.

This solves the package:test use case, but not the general `dart <file>` use case. Maybe that is a separate problem but they seem very related to me.

**Comment by johnniwinther** [comment]
kevmoo@ the amount of work needed in the CFE to support detecting the opt-in state from the entry-point is substantial: +1 week of work to make it any better than just compiling twice if we guessed the wrong mode on the first try.

**Comment by natebosch** [comment]
> I am not sure I understand how the tools detect the mode in order to pass it down to CFE.

The same way we'd detect the mode for picking how to interpret the syntax of the file. @lrhn might be able to clear up any confusion on the details. It should involve reading the `package_config.json` file, picking which package the library is in, and then scanning the library for a language version comment.

> better than just compiling twice if we guessed the wrong mode on the first try.

This would probably be fine. I can't imagine the runtime hit would be overwhelming, and as the migration continues the chances of getting it wrong drop further and further.

**Comment by natebosch** [comment]
I should add, though, that if we choose the compile-then-retry method we would need to consider a non-migrated _entrypoint_ as OK to retry, but a non-migrated _transitive_ import as a failure.

**Comment by lrhn** [comment]
Re. 

> * picking the default version based on the package: of the library spawning the URI.

I want to very much discourage that approach. It basically requires looking up the stack to find the function which called `Isolate.spawnUri`, then figure out its library and package. That completely breaks any attempt of abstraction, like using `package:isolate`.

**Comment by kevmoo** [comment]
Had a discussion with @jakemac53 @natebosch @johnniwinther and @leafpetersen 

We have a plan.  Will open a new issue here under https://github.com/dart-lang/sdk/issues/41203
